core/src/main/java/fr/sncf/osrd/standalone_sim/SignalProjection.kt

```diff
@@ -0,0 +1,262 @@
+package fr.sncf.osrd.standalone_sim
+
+import fr.sncf.osrd.api.FullInfra
+import fr.sncf.osrd.api.SignalProjectionEndpoint
+import fr.sncf.osrd.api.SignalProjectionEndpoint.SignalProjectionResult
+import fr.sncf.osrd.envelope.Envelope
+import fr.sncf.osrd.infra_state.api.TrainPath
+import fr.sncf.osrd.signaling.SignalingSimulator
+import fr.sncf.osrd.signaling.ZoneStatus
+import fr.sncf.osrd.sim_infra.api.*
+import fr.sncf.osrd.sim_infra_adapter.SimInfraAdapter
+import fr.sncf.osrd.standalone_sim.result.ResultTrain.SignalSighting
+import fr.sncf.osrd.standalone_sim.result.ResultTrain.ZoneUpdate
+import fr.sncf.osrd.standalone_sim.result.SignalUpdate
+import fr.sncf.osrd.utils.indexing.MutableStaticIdxArrayList
+import fr.sncf.osrd.utils.indexing.StaticIdxList
+import fr.sncf.osrd.utils.indexing.mutableStaticIdxArrayListOf
+import java.awt.Color
+
+data class SignalAspectChangeEvent(val newAspect: String, val time: Long, val offset: Distance)
+
+fun project(
+    fullInfra: FullInfra,
+    trainPath: TrainPath,
+    signalSightings: List<SignalSighting>,
+    zoneUpdates: List<ZoneUpdate>
+): SignalProjectionResult {
+    val rawInfra = fullInfra.rawInfra;
+    val loadedSignalInfra = fullInfra.loadedSignalInfra;
+    val blockInfra = fullInfra.blockInfra;
+    val simulator = fullInfra.signalingSimulator;
+
+    // TODO: allowed signaling systems should depend on the type of train
+    val sigSystemManager = simulator.sigModuleManager
+    val bal = sigSystemManager.findSignalingSystem("BAL")
+    val bapr = sigSystemManager.findSignalingSystem("BAPR")
+    val tvm = sigSystemManager.findSignalingSystem("TVM")
+
+    // Get the route path
+    // TODO: do it in the pathfinding
+    val routes = MutableStaticIdxArrayList<Route>()
+    for (javaRoute in trainPath.routePath) {
+        val route = rawInfra.routeMap[javaRoute.element.infraRoute]!!
+        routes.add(route)
+    }
+
+    val blockPaths = getRouteBlocks(
+        rawInfra, blockInfra, routes, mutableStaticIdxArrayListOf(bal, bapr, tvm)
+    )
+    assert(blockPaths.isNotEmpty())
+    val blockPath = blockPaths[0] // TODO: have a better way to choose the block path
+
+    val zoneMap = mutableMapOf<String, Int>()
+    var zoneCount = 0
+    for (block in blockPath) {
+        for (zonePath in blockInfra.getBlockPath(block)) {
+            val zone = rawInfra.getNextZone(rawInfra.getZonePathEntry(zonePath))!!
+            val zoneName = rawInfra.getZoneName(zone)
+            zoneMap.put(zoneName, zoneCount)
```

Comment:
Intellij says `map.put() should be converted to assignment `

I'm not familiar enough with kotlin to know if that makes sense

Reply:
It does. (Intellij is really good at kotlin, which is fitting since it's a jetbrains language)